File: P3/Civitas/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;
import java.util.Collections;
/**
 *
 * @author salva
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    MazoSorpresas(){
        init();
        debug = false;
    }
    
    MazoSorpresas(boolean d){
        init();
        debug = d;
        if (d)
            Diario.getInstance().ocurreEvento("debug mazo: true");
    }
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    void alMazo(Sorpresa s){
        if (!barajada){
            sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente(){
        if (!barajada || usadas == 0){
            usadas = 0;
            barajada = true;
            if (!debug)
                Collections.shuffle(sorpresas);
        }
        usadas += 1;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        boolean estaEnMazo = sorpresas.contains(sorpresa);
        if (estaEnMazo){
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se inhabilita carta sorpresa "+sorpresa.getNom());
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        boolean estaEnMazo = cartasEspeciales.contains(sorpresa);
        if (estaEnMazo){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se habilitad carta sorpresa "+ sorpresa.getNom());  
        }
    }
}

File: ExamenParcial1/civitasruby/lib/civitas/mazo_sorpresas.rb
#encoding:utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
#require_relative "sorpresa"
require_relative "diario"
module Civitas
  class Mazo_sorpresas
    def initialize(d=false)
      init
      @ultima_sorpresa
      @debug = d
      if @debug
        Diario.instance.ocurre_evento("Activado modo debug MazoSorpresas")
      end
    end

    def init
      @sorpresas = Array.new
      @cartas_especiales = Array.new
      @barajada = false
      @usadas = 0
    end

    def al_mazo(sorpresa)
      if not @barajada
        @sorpresas << sorpresa
      end
    end

    def siguiente
      if not @barajada or @usadas == @sorpresas.length
        @usadas = 0
        if not @debug
          @sorpresas.shuffle
          @barajada = true
        end
      end

      @usadas += 1
      @ultima_sorpresa= @sorpresas.shift
      @sorpresas << @ultima_sorpresa
      return @ultima_sorpresa
    end

    def inhabilitar_carta_especial(sorpresa)
      if @sorpresas.include?(sorpresa)
        @sorpresas.delete(sorpresa)
        @cartas_especiales << sorpresa
        Diario.instance.ocurre_evento("Inhabilitada carta especial")
      end
    end

    def habilitar_carta_especial(sorpresa)
      if @cartas_especiales.include?(sorpresa)
        @cartas_especiales.delete(sorpresa)
        @sorpresas << sorpresa
        Diario.instance.ocurre_evento("Se habilita la sorpresa " + sorpresa.to_s)
      end
    end

    private :init
  end
end
